---
title: "16S Practice"
author: "Nathan"
date: "12/27/2022"
output: html_document
---

**This follows along with the [DADA2 Pipeline Tutorial](https://benjjneb.github.io/dada2/tutorial.html) (1.16) online .**

Here we walk through version 1.16 of the DADA2 pipeline on a small multi-sample dataset. Our starting point is a set of Illumina-sequenced paired-end fastq files that have been split (or "demultiplexed") by sample and from which the barcodes/adapters have already been removed. The end product is an **amplicon sequence variant (ASV) table**, a higher-resolution analogue of the traditional OTU table, which records the number of times each [exact amplicon sequence variant](https://www.nature.com/articles/ismej2017119) was observed in each sample. We also assign taxonomy to the output sequences, and demonstrate how the data can be imported into the popular [phyloseq](https://joey711.github.io/phyloseq/) R package for the analysis of microbiome data.

<!--# Is there a way i can upload this to Github? -->

<!--# DO WE NEED TO MAKE THIS OTU TO BE LESS SPECIFIC? -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries

```{r}
library(dada2); packageVersion("dada2")
library(dplyr)
library(ggplot2)
library(tidyverse)
library(here)
library(lubridate)

```

# Getting Ready

#### Set Path

Define the following path variable so that it points to the extracted directory on **your** machine:

```{r}
path <- "/Users/nathangreenslit/Desktop/UGA/Lipp Lab/Sequencing/code/wrangling/DADA2/16S_TX_FastQ" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)
```

<!--# I deleted NTC and DI Blanks -->

#### Group Files by Forward or Reverse Reads

Now we read in the names of the fastq files, and perform some string manipulation to get matched lists of the forward and reverse fastq files.

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

# Inspecting Read Quality

In gray-scale is a heat map of the frequency of each quality score at each base position. The mean quality score at each position is shown by the green line, and the quartiles of the quality score distribution by the orange lines. The red line shows the scaled proportion of reads that extend to at least that position (this is more useful for other sequencing technologies, as Illumina reads are typically all the same length, hence the flat red line).

We generally advise trimming the last few nucleotides to avoid less well-controlled errors that can arise there.

#### Forward Reads

```{r}
plotQualityProfile(fnFs[1:2])
```

#### Reverse Reads

```{r}
plotQualityProfile(fnRs[1:2])
```

Trimming as the average qualities crash will improve the algorithm's sensitivity to rare sequence variants.

Based on these profiles, <!--# we will truncate forwards at 250 and R at 200?? -->

**Considerations for your own data:** **Your reads must still overlap after truncation in order to merge them later!** The tutorial is using 2x250 V4 sequence data, so the forward and reverse reads almost completely overlap and our trimming can be completely guided by the quality scores.

If you are using a less-overlapping primer set, like V1-V2 or V3-V4, your `truncLen` must be large enough to maintain `20 + biological.length.variation` nucleotides of overlap between them.

<!--# We are using V4 sequence data -->

# Filter and Trim

#### Assign filenames for the filtered fastq.gz files

```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

#### Filter and Trim where quality starts to drop

We'll use standard filtering parameters: `maxN=0` (DADA2 requires no Ns), `truncQ=2`, `rm.phix=TRUE` and `maxEE=2`. The `maxEE` parameter sets the maximum number of "expected errors" allowed in a read, which is [a better filter than simply averaging quality scores](https://academic.oup.com/bioinformatics/article/31/21/3476/194979).

```{r}
out <- filterAndTrim(
  fnFs, filtFs, fnRs, filtRs, truncLen=c(250,200), #insert where we trim for F/R
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

**Considerations for your own data:** The standard filtering parameters are starting points, not set in stone. If you want to speed up downstream computation, consider tightening `maxEE`. If too few reads are passing the filter, consider relaxing `maxEE`, perhaps especially on the reverse reads (eg. `maxEE=c(2,5)`), and reducing the `truncLen` to remove low quality tails. Remember though, when choosing `truncLen` for paired-end reads you must maintain overlap after truncation in order to merge them later.

<!--# Should I loosen maxEE to allow for more reads? -->

# Learn Error Rates

The DADA2 algorithm makes use of a parametric error model (`err`) and every amplicon dataset has a different set of error rates. The `learnErrors` method learns this error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution. As in many machine-learning problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).

#### Learn error rates for Forward/Reverse Reads

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
```

```{r}
errR <- learnErrors(filtRs, multithread=TRUE)
```

#### Plot error rates

```{r}
plotErrors(errF, nominalQ=TRUE)

```

The error rates for each possible transition (A→C, A→G, ...) are shown. **Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score.**

We want the estimated error rates [(black line) to be a good fit to the observed rates (points), and the error rates to drop with increased quality as expected]{.ul}. Everything looks reasonable and we proceed with confidence.

# Sample Inference

#### Apply the Sample Inference Algorithm to the filtered and trimmed sequence data

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)

```

```{r}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

```

#### Look at how many true variants from the total amount of sequences in Sample 1

```{r}
dadaFs[[1]]

```

The DADA2 algorithm inferred 36 true sequence variants from the 418 unique sequences in the first sample. There is much more to the `dada-class` return object than this (see `help("dada-class")` for some info), including multiple diagnostics about the quality of each denoised sequence variant, but that is beyond the scope of an introductory tutorial.

# Merging Paired Reads

We now merge the forward and reverse reads together to obtain the full denoised sequences. Merging is performed by aligning the denoised forward reads with the reverse-complement of the corresponding denoised reverse reads, and then constructing the merged "contig" sequences. By default, merged sequences are only output if the forward and reverse reads overlap by at least 12 bases, and are identical to each other in the overlap region (but these conditions can be changed via function arguments).

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1 ]])
```

The `mergers` object is a list of `data.frame`s from each sample. Each `data.frame` contains the merged `$sequence`, its `$abundance`, and the indices of the `$forward` and `$reverse` sequence variants that were merged. Paired reads that did not exactly overlap were removed by `mergePairs`, further reducing spurious output.

**Considerations for your own data:** Most of your **reads** should successfully merge. If that is not the case upstream parameters may need to be revisited: Did you trim away the overlap between your reads?

# Construct Sequence Table

We can now construct an amplicon sequence variant table (ASV) table, a higher-resolution version of the OTU table produced by traditional methods.

<!--# Do i need to make this OTU? -->

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

```{r}
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

The sequence table is a `matrix` with rows corresponding to (and named by) the samples, and columns corresponding to (and named by) the sequence variants. This table contains 1593 ASVs, and the lengths of our merged sequences all fall within the expected range for this V4 amplicon.

<!--# Is this true for V4? = 291bp... -->

**Considerations for your own data:** Sequences that are much longer or shorter than expected may be the result of non-specific priming. You can remove non-target-length sequences from your sequence table (eg. `seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% 250:256]`). This is analogous to "cutting a band" in-silico to get amplicons of the targeted length.

<!--# May need to do this -->

# Remove Chimeras

The core `dada` method corrects substitution and indel errors, but chimeras remain. Fortunately, the accuracy of sequence variants after denoising makes identifying chimeric ASVs simpler than when dealing with fuzzy OTUs. Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant "parent" sequences.

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

The frequency of chimeric sequences varies substantially from dataset to dataset, and depends on on factors including experimental procedures and sample complexity. Here chimeras make up about X% of the merged sequence variants, but when we account for the abundances of those variants we see they account for about 20% of the merged sequence reads.

<!--# Is this a lot? -->

**Considerations for your own data:** Most of your **reads** should remain after chimera removal (it is not uncommon for a majority of **sequence variants** to be removed though). If most of your reads were removed as chimeric, upstream processing may need to be revisited. In almost all cases this is caused by primer sequences with ambiguous nucleotides that were not removed prior to beginning the DADA2 pipeline.

# Track reads through the pipeline

As a final check of our progress, we'll look at the number of reads that made it through each step in the pipeline:

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
# rownames(track) <- sample.names
head(track)
```

Large drops in any step are undesired.

<!--# What classifies a large drop? -->

**Considerations for your own data:** This is a great place to do a last **sanity check**. Outside of filtering, there should no step in which a majority of reads are lost. If a majority of reads failed to merge, you may need to revisit the `truncLen` parameter used in the filtering step and make sure that the truncated reads span your amplicon. If a majority of reads were removed as chimeric, you may need to revisit the removal of primers, as the ambiguous nucleotides in unremoved primers interfere with chimera identification

# Assign Taxonomy

It is common at this point, especially in 16S/18S/ITS amplicon sequencing, to assign taxonomy to the sequence variants. The DADA2 package provides a native implementation of the [naive Bayesian classifier method](http://www.ncbi.nlm.nih.gov/pubmed/17586664) for this purpose. The `assignTaxonomy` function takes as input a set of sequences to be classified and a training set of reference sequences with known taxonomy, and outputs taxonomic assignments with at least `minBoot` bootstrap confidence.

We maintain [formatted training fastas for the RDP training set, GreenGenes clustered at 97% identity, and the Silva reference database](https://benjjneb.github.io/dada2/training.html), and additional trainings fastas suitable for protists and certain specific environments have been contributed. For fungal taxonomy, the General Fasta release files from the [UNITE ITS database](https://unite.ut.ee/repository.php) can be used as is. To follow along, download the `silva_nr_v132_train_set.fa.gz` file, and place it in the directory with the fastq files.

#### Assigning Taxonomy from Silva v132 database

Reference database is stored in separate folder from Project so that I can upload to Github (the ref db is \>100mb)

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/nathangreenslit/Desktop/UGA/R/r_wd/Sequencing/silva_nr_v132_train_set.fa.gz", multithread=TRUE)
```

**Extensions:** The dada2 package also implements a method to make [species level assignments based on **exact matching**](https://benjjneb.github.io/dada2/assign.html#species-assignment) between ASVs and sequenced reference strains. Recent analysis suggests that [exact matching (or 100% identity) is the only appropriate way to assign species to 16S gene fragments](https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/bty113/4913809). Currently, [species-assignment training fastas are available for the Silva and RDP 16S databases](https://benjjneb.github.io/dada2/training.html). To follow the optional species addition step, download the `silva_species_assignment_v132.fa.gz` file, and place it in the directory with the fastq files.

    taxa <- addSpecies(taxa, "~/tax/silva_species_assignment_v132.fa.gz")

<!--# Should i do this? -->

#### Inspect the taxonomic assignments

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

**Considerations for your own data:** If your reads do not seem to be appropriately assigned, for example lots of your bacterial 16S sequences are being assigned as `Eukaryota NA NA NA NA NA`, your reads may be in the opposite orientation as the reference database. Tell dada2 to try the reverse-complement orientation with `assignTaxonomy(..., tryRC=TRUE)` and see if this fixes the assignments. If using DECIPHER for taxonomy, try `IdTaxa (..., strand="both")`.

**Alternatives:** The recently developed `IdTaxa` taxonomic classification method is also available via [the DECIPHER Bioconductor package](https://bioconductor.org/packages/release/bioc/html/DECIPHER.html). The paper introducing the IDTAXA algorithm [reports classification performance that is better than the long-time standard set by the naive Bayesian classifier](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0521-5). Here we include a code block that allows you to use `IdTaxa` as a drop-in replacement for `assignTaxonomy` (and it's faster as well!). Trained classifiers are available from [\<http://DECIPHER.codes/Downloads.html>](http://decipher.codes/Downloads.html). Download the SILVA SSU r132 (modified) file to follow along.

    library(DECIPHER); packageVersion("DECIPHER")

    ## [1] '2.14.0'

    dna <- DNAStringSet(getSequences(seqtab.nochim)) # Create a DNAStringSet from the ASVs
    load("~/tax/IDTaxa/SILVA_SSU_r132_March2018.RData") # CHANGE TO THE PATH OF YOUR TRAINING SET
    ids <- IdTaxa(dna, trainingSet, strand="top", processors=NULL, verbose=FALSE) # use all processors
    ranks <- c("domain", "phylum", "class", "order", "family", "genus", "species") # ranks of interest
    # Convert the output object of class "Taxa" to a matrix analogous to the output from assignTaxonomy
    taxid <- t(sapply(ids, function(x) {
            m <- match(ranks, x$rank)
            taxa <- x$taxon[m]
            taxa[startsWith(taxa, "unclassified_")] <- NA
            taxa
    }))
    colnames(taxid) <- ranks; rownames(taxid) <- getSequences(seqtab.nochim)

The `taxid` matrix from `IdTaxa` is a drop-in replacement for the `taxa` matrix from `assignTaxonomy`, simply set `taxa <- taxid` to carry on using the `IdTaxa` assignments. Read more about `IdTaxa` and other useful tools at [the DECIPHER R package website](http://www2.decipher.codes/).

 

# Evaluate Accuracy

One of the samples included was a "mock community (ZYMO Positive Control)", in which a mixture of known strains was sequenced. We return to that sample and compare the sequence variants inferred by DADA2 to the expected composition of the community.

<!--# Revisit this for my positive control -->

```{r}
#{r} unqs.mock <- seqtab.nochim["Mock",] unqs.mock <- sort(unqs.mock[unqs.mock>0], decreasing=TRUE) # Drop ASVs absent in the Mock cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")
```

```{r}
#{r} mock.ref <- getSequences(file.path(path, "HMP_MOCK.v35.fasta")) match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref)))) cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

# Hand off to Phyloseq

```{r}
library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(microViz)
theme_set(theme_bw())
```

#### Make my own dataframe

```{r}
samples.out <- rownames(seqtab.nochim)
subject <- sapply(strsplit(samples.out, "NONE"), `[`, 1)

#Site Column:
gender <- substr(subject,8,9) 
#8 and 9 are the character numbers. Basically going from 264-NG-BO-19-5437 and pulling the 8th and 9th characters "B" and "O"

#Date Column:
subject <- substr(subject,11,12)  #Same thing here for date "1" and "2" = 12th 
day <- as.integer(sapply(strsplit(samples.out, "D"), `[`, 2)) #Does something with the name..
samdf <- data.frame(Date=subject, Site=gender, Day=day) #Renaming gender and subject as Date and Site

samdf$When <- "Early"
samdf$When[samdf$Day>100] <- "Late"
rownames(samdf) <- samples.out
```

#### Construct a Phlyoseq object

```{r}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps <- prune_samples(sample_names(ps) != "Mock", ps) # Remove mock sample. 
```

<!--# Will need to update Mock to Positive Control -->

It is more convenient to use short names for our ASVs (e.g. ASV21) rather than the full DNA sequence when working with some of the tables and visualizations from phyloseq, but we want to keep the full DNA sequences for other purposes like merging with other datasets or indexing into reference databases like [the Earth Microbiome Project](http://www.earthmicrobiome.org/).

For that reason we'll store the DNA sequences of our ASVs in the refseq slot of the phyloseq object, and then rename our taxa to a short string.

That way, the short new taxa names will appear in tables and plots, and we can still recover the DNA sequences corresponding to each ASV as needed with `refseq(ps)`.

```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps)) #Maybe gets specific nice names for taxa
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna) #Merge these two 
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

#### Visualize Alpha Diveristy

<!--# Maybe do this is visual code with a df object rather than phyloseq object since phyloseq sucks -->

<!--# I am thinking we melt here and then do all these plots in the visualization code.  -->

[**Shannon's index**]{.ul}: believed to emphasize the richness component of diversity

[**Simpson's index**]{.ul}: emphasizing the evenness component

```{r}
plot_richness(ps, x="Date", measures=c("Shannon", "Simpson"), color="Site")
```

#### Ordinate

```{r}
# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(ps, function(otu) otu/sum(otu))
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")
```

#### Plot ordination

```{r}
plot_ordination(ps.prop, ord.nmds.bray, color="Site", title="Bray NMDS")
```

<!--# What does Bray Ordination Do? -->

#### Bar Plot

```{r, fig.width= 10}
top300 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:100000] #Selects top N to look at
ps.top300 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top300 <- prune_taxa(top300, ps.top300)
```

#Make a dataframe so we can work with it

```{r}
ps_1<- psmelt(ps.top300) %>%
 filter(!Order %in% "Chloroplast") %>% #Gets rid of chloroplasts
mutate(ddate = case_when( Date == "7-" ~ "2022-07-07",
                          Date == "8-" ~ "2022-07-08",
                          Date == "9-" ~ "2022-07-09",
                          Date == "11" ~ "2022-07-11",
                          Date == "12" ~ "2022-07-12",
                          Date == "13" ~ "2022-07-13",
                          Date == "14" ~ "2022-07-14",
                          Date == "15" ~ "2022-07-15",
                          Date == "16" ~ "2022-07-16",
                          Date == "17" ~ "2022-07-17",
                          Date == "18" ~ "2022-07-18",
                          Date == "19" ~ "2022-07-19")) 
ps_1<- ps_1 %>%
  mutate(ddate = as.Date(ps_1$ddate))
```

Save as RDS

```{r}
saveRDS(ps_1, file = here("data", "clean_data", "clean_16S_practice.rds"))

saveRDS(ps, file = here("data", "clean_data", "clean_16s_phyloseq_practice.rds"))
```
